ls1intum/Artemis
# Re-evaluate automatic results of programming exercises

### Checklist

### Motivation and Context
Currently, building all participations (e.g. on Bamboo) is needed if e.g. the test case weights are changed. There should be a less expensive option.

### Description
- Added database query to get all latest automatic results of a programming exercise
- Added secured REST endpoint `/programming-exercise/{exerciseId}/re-evaluate`
- Added a new `updateAllResultsFromTestCases` method that returns updated versions of all the latest automatic results.
- Fixed a possible security problem (missing access check) for `updateTestCases`

### Steps for Testing
1. Create a programming exercise
2. Participate (potentially multiple times with different students) with some passing and some failing tests
3. Navigate into `Manage Test Cases`
4. Change several entries (e.g. weight, bonus multiplier, bonus points) and click on `Save test cases`
5. Click on `Re-evaluate all` (see screenshot below)
6. Check that the results in the `Scores` page change (without the need to build those results again on Bamboo)

### Screenshots
![image](https://user-images.githubusercontent.com/744067/89632292-3ee94d00-d8a2-11ea-9e6f-7871f20e5b65.png)
